### PR TITLE
Fix docker swarm mode

### LINKED ISSUE #216: Issue with Swarm Mode V2
While testing for resolving #178, i found the reason why there is a issue with the services created by ouroboros:

If a service was updated with the docker cli, the inspect of a container shows this image config:
`"Image": "nginx:stable-alpine@sha256:b96aeeb1687703c49096f4969358d44f8520b671da94848309a3ba5be5b4c632",`

With this config, ouroboros wants to update the container but there is no newer image.
If you run the update, it replace the container with the same image.

The same container created by ouroboros have this image config:
`"Image": "nginx:stable-alpine",`

After that, ouroboros skip the containers because it is not able to find the SHA of the image.
The needed SHA is in both cases shown as part of the inspect, but not as a part of the config.

`"Image": "sha256:66952fd0a8efa0598626fad89d3a0827bc24fc92c3adb576adbc9fd58606e1af",`

I added both inspects to Gist:
https://gist.github.com/computeralex92/2f93cd64656850b2d2d08970c6d21b1a
https://gist.github.com/computeralex92/43ea1953d256324ef7caa0e5711df9bf

### PR DESCRIPTION
**Pull request by mathcantin**
Fixes #216 

The fix is instead of using service.update(image=tag) we need to use service.update(image=f"{tag}@sha256:{latest_image_sha256}").

With a little bit of refractor to remove duplicate code.

### CODE REVIEW

**Comment by mathcantin** [comment]
On the issues, there are this comment
> as another way to look at it he is saying it tries to update something that doesnt need updating, which is odd since its checking the sha when it DOES exist initially. Maybe the root can be found there

Still a good idea, but that not cover by this PR.

**Comment by DirtyCajunRice** [comment]
have you ran this code? i like the baseobject but a lot of stuff just wont work. e.g. digest or something doesnt return digest or something. it returns true or false

**Comment by mathcantin** [comment]
Yes, I try with and without swarm mode and that works. The method _get_digest, return the sha256 of the image, in all cases (dryrun on or off with auth or not). I don't know where you see a method that return a boolean.

**Comment by mathcantin** [comment]
I understand, the code:
```
        digest = image.attrs.get("Descriptor", {}).get("digest")
        digest = digest or image.attrs.get("RepoDigests")[0].split('@')[1]
        digest = digest or image.id

        return self._remove_sha_prefix(digest)
```

is exactly the same thing then that

``` 
    digest = image.attrs.get("Descriptor", {}).get("digest")
    if not digest:
            digest = image.attrs.get("RepoDigests")[0].split('@')[1]
    if not digest
        digest = image.id

        return self._remove_sha_prefix(digest)
```

Because the or keyword isn't a comparator and don't return True or False. I can change it, if you prefer the second solution.

**Comment by DirtyCajunRice** [comment]
Yeah i spoke before i ran the code myself. its completely fine. Im reading over the change right now. Give me a bit is a large change.

File: pyouroboros/dockerclient.py [hunk 1]
```diff
@@ -381,38 +383,24 @@ def monitor_filter(self):
...
-                    exit(1)
-            elif 'Client.Timeout' in str(e):
-                self.logger.critical("Couldn't find an image on docker.com for %s. Local Build?", tag)
-                raise ConnectionError
-            elif ('pull access' or 'TLS handshake') in str(e):
-                self.logger.critical("Couldn't pull. Skipping. Error: %s", e)
-                raise ConnectionError
+        return self._pull(tag)
+
+    def _remove_sha_prefix(self, digest):
```
Comment (DirtyCajunRice) [inline pyouroboros/dockerclient.py]: This is only called in one place, so why is it abstracted to its own private function?

Comment (mathcantin) [reply pyouroboros/dockerclient.py]: I use it two times. That improve readability of code.

**Pull request merged**